File: taxonomy.py
"""
taxonomy.py — the single source of truth for the supply-chain vocabulary.

The project used to model each chain as flow[] -> tier -> players[] using a fixed
bottom-up 10-tier list. That has been replaced by the user's "AI Build Supply Chain"
taxonomy: a TOP-DOWN stack of LAYERS (Application at the top, Critical Minerals at the
bottom), where each layer holds SECTORS (and a few sectors hold SUB-SECTORS), plus four
cross-cutting DOMAINS (power, thermal, security, edge AI) that sit beside the stack.

Everything that needs to know "what are the layers / what color is each / what order do
they go in" imports from here, so the vocabulary lives in ONE place instead of being
copy-pasted across app.py, graph_build.py and main.py.

Data files store SLUGS (the short machine names below) for `layer` and `domain`, and
free strings for `sector` / `sub_sector`. Human-readable display names come from the
tables here.
"""

# ── The 13 layers, TOP -> BOTTOM (this ORDER is canonical: it drives column order) ──
# Each row is (slug, display name, hex color).
# "system_integration" is a deliberate addition by the project owner — it is NOT in the
# source taxonomy PDF (which has 12 layers). It is the home for server OEM/ODM
# integrators (Foxconn, Quanta, Wistron, Supermicro, Dell, Arista), which the 12-layer
# stack has no clean slot for. It sits between Cloud Infrastructure and Compute Hardware.
LAYERS = [
    ("application",        "Application",              "#a855f7"),
    ("ai_models",          "AI Models",                "#8b5cf6"),
    ("software_infra",     "Software Infrastructure",  "#6366f1"),
    ("cloud_infra",        "Cloud Infrastructure",     "#3b82f6"),
    ("system_integration", "System Integration",       "#16a34a"),  # owner-added (not in PDF)
    ("compute_hardware",   "Compute Hardware",         "#0ea5e9"),
    ("memory",             "Memory",                   "#06b6d4"),
    ("interconnect",       "Interconnect",             "#14b8a6"),
    ("advanced_packaging", "Advanced Packaging",       "#ec4899"),
    ("foundry",            "Semiconductor Foundry",    "#f97316"),
    ("equipment",          "Semiconductor Equipment",  "#f59e0b"),
    ("materials",          "Semiconductor Materials",  "#84cc16"),
    ("minerals",           "Critical Minerals",        "#a16207"),
]

# ── The 4 cross-cutting domains (drawn beside the stack, not inside it) ──────────────
DOMAINS = [
    ("power",    "Power Infrastructure", "#dc2626"),
    ("thermal",  "Thermal Management",   "#0891b2"),
    ("security", "Security",             "#7c3aed"),
    ("edge_ai",  "Edge & Physical AI",   "#db2777"),
]

# ── Derived lookups (built once, imported everywhere) ───────────────────────────────
LAYER_ORDER  = {slug: i for i, (slug, *_) in enumerate(LAYERS)}        # for sorting columns
LAYER_COLORS = {slug: color for slug, _name, color in LAYERS}
DOMAIN_COLORS = {slug: color for slug, _name, color in DOMAINS}
LAYER_NAMES  = {slug: name for slug, name, _color in LAYERS}           # display in headers/badges
DOMAIN_NAMES = {slug: name for slug, name, _color in DOMAINS}

# Convenience: one color map covering both layers and domains (layers win on overlap,
# but the slug sets are disjoint so order doesn't matter). Handy for node coloring where
# a node may be placed in either a layer or a domain.
GROUP_COLORS = {**LAYER_COLORS, **DOMAIN_COLORS}
GROUP_NAMES  = {**LAYER_NAMES, **DOMAIN_NAMES}

# Valid slug sets — used by graph_build.py / tools to catch typos in chain files.
LAYER_SLUGS  = set(LAYER_COLORS)
DOMAIN_SLUGS = set(DOMAIN_COLORS)


def iter_players(chain):
    """Walk every player in a chain file, regardless of nesting depth.

    A chain has an optional `flow` (the layer stack) and an optional `domains` block.
    Both have the same shape:  group -> sectors[] -> (sub_sectors[]) -> players[].
    A sector holds EITHER `players` directly OR `sub_sectors` (each with `players`).

    Yields a 5-tuple for every player:
        (player_dict, group_slug, kind, sector, sub_sector)
    where:
        group_slug  is the layer slug (kind == "layer") or domain slug (kind == "domain")
        kind        is "layer" or "domain"
        sector      is the sector string (or None if the group has no sectors)
        sub_sector  is the sub-sector string, or None

    This is the ONE place the nested loop lives, so callers (graph_build.py, app.py)
    never re-implement the 4-deep walk.
    """
    groups = (
        (chain.get("flow", []),    "layer"),
        (chain.get("domains", []), "domain"),
    )
    for group_list, kind in groups:
        slug_key = "layer" if kind == "layer" else "domain"
        for group in group_list:
            group_slug = group.get(slug_key)
            for sector in group.get("sectors", []):
                sec_name = sector.get("sector")
                # sub-sectors first (e.g. Interconnect > Scale-up > CPO)
                for sub in sector.get("sub_sectors", []):
                    sub_name = sub.get("sub_sector")
                    for player in sub.get("players", []):
                        yield player, group_slug, kind, sec_name, sub_name
                # players attached directly to the sector
                for player in sector.get("players", []):
                    yield player, group_slug, kind, sec_name, None

File: web/src/lib/taxonomy.ts
// taxonomy.ts — port of taxonomy.py, the single source of truth for the
// supply-chain vocabulary (layers, domains, colors, display names, order).
// Data files store SLUGS for `layer`/`domain`; display names come from here.

type Row = [slug: string, name: string, color: string];

// The 13 layers, TOP -> BOTTOM. This ORDER is canonical and drives column order.
export const LAYERS: Row[] = [
  ["application", "Application", "#a855f7"],
  ["ai_models", "AI Models", "#8b5cf6"],
  ["software_infra", "Software Infrastructure", "#6366f1"],
  ["cloud_infra", "Cloud Infrastructure", "#3b82f6"],
  ["system_integration", "System Integration", "#16a34a"],
  ["compute_hardware", "Compute Hardware", "#0ea5e9"],
  ["memory", "Memory", "#06b6d4"],
  ["interconnect", "Interconnect", "#14b8a6"],
  ["advanced_packaging", "Advanced Packaging", "#ec4899"],
  ["foundry", "Semiconductor Foundry", "#f97316"],
  ["equipment", "Semiconductor Equipment", "#f59e0b"],
  ["materials", "Semiconductor Materials", "#84cc16"],
  ["minerals", "Critical Minerals", "#a16207"],
];

// The 4 cross-cutting domains (drawn beside the stack, not inside it).
export const DOMAINS: Row[] = [
  ["power", "Power Infrastructure", "#dc2626"],
  ["thermal", "Thermal Management", "#0891b2"],
  ["security", "Security", "#7c3aed"],
  ["edge_ai", "Edge & Physical AI", "#db2777"],
];

// Derived lookups.
export const LAYER_ORDER: Record<string, number> = Object.fromEntries(
  LAYERS.map(([slug], i) => [slug, i])
);
export const LAYER_COLORS: Record<string, string> = Object.fromEntries(
  LAYERS.map(([slug, , color]) => [slug, color])
);
export const DOMAIN_COLORS: Record<string, string> = Object.fromEntries(
  DOMAINS.map(([slug, , color]) => [slug, color])
);
export const LAYER_NAMES: Record<string, string> = Object.fromEntries(
  LAYERS.map(([slug, name]) => [slug, name])
);
export const DOMAIN_NAMES: Record<string, string> = Object.fromEntries(
  DOMAINS.map(([slug, name]) => [slug, name])
);

// One color/name map covering both layers and domains (slug sets are disjoint).
export const GROUP_COLORS: Record<string, string> = { ...LAYER_COLORS, ...DOMAIN_COLORS };
export const GROUP_NAMES: Record<string, string> = { ...LAYER_NAMES, ...DOMAIN_NAMES };

// Domain order index (for Chain 2D right-panel ordering).
export const DOMAIN_ORDER: Record<string, number> = Object.fromEntries(
  DOMAINS.map(([slug], i) => [slug, i])
);

// Per-chain edge colors (ported from app.py CHAIN_COLORS). Edges are colored by chain.
export const CHAIN_COLORS: Record<string, string> = {
  nvidia_vera_rubin: "#76b900",
  google_tpu_v7_ironwood: "#4285f4",
  nvda_b200: "#ff6b35",
  optical_networking: "#8b5cf6",
  broadcom_custom_asic: "#f43f5e",
  hbm_memory: "#06d6a0",
  nand_flash: "#f59e0b",
  cpu_datacenter: "#ef4444",
  mlcc: "#0ea5e9",
  power_cooling: "#f97316",
  foundry: "#a78bfa",
  amd_mi450_helios: "#ed1c24",
  aws_trainium2: "#ff9900",
  packaging_substrate: "#14b8a6",
  tpu_v8t: "#1a73e8",
  tpu_v8i: "#34a853",
  amd_mi355: "#b91c1c",
  aws_trainium3: "#cc7a00",
  neocloud: "#2dd4bf",
  power_semiconductor: "#eab308",
};

export const groupColor = (slug: string): string => GROUP_COLORS[slug] || "#94a3b8";
export const groupName = (slug: string): string => GROUP_NAMES[slug] || slug;
export const chainColor = (slug: string): string => CHAIN_COLORS[slug] || "#8b949e";

// Turn a chain slug into a readable label ("nvda_b200" -> "nvda b200").
export const slugLabel = (slug: string): string => slug.replace(/_/g, " ");
